### PR TITLE
ORC-1352: [C++] Remove `ORC_[NOEXCEPT|NULLPTR|OVERRIDE|UNIQUE_PTR]` macro usages

### PR DESCRIPTION
**Pull request by zhjwpku**
### What changes were proposed in this pull request?

the following macros could be removed from the main source code to make the code neat.

```
#define ORC_NOEXCEPT noexcept
#define ORC_NULLPTR nullptr
#define ORC_OVERRIDE override
#define ORC_UNIQUE_PTR std::unique_ptr
```

These macros should only be keeped in the `orc-config.hh.in` for backward compatiblity(in case any user use these marcros).

### Why are the changes needed?

to make the code more readable.

### How was this patch tested?

It doesn't introduce new features and passed all test cases.

### CODE REVIEW

**Review by wgtmac — APPROVED** [review]
LGTM. cc @stiga-huang

File: c++/include/orc/orc-config.hh.in [hunk 1]
```diff
@@ -25,6 +25,7 @@
   #include <stdint.h>
 #endif
 
+// Following MACROS should be keeped for backward compatibility.
```
Comment (dongjoon-hyun) [inline c++/include/orc/orc-config.hh.in:28]: Just a question. Do we have any deprecation process in Apache ORC C++, @wgtmac , @stiga-huang ?
I'm wondering when we can clean up these unused macros in the future.

**Comment by dongjoon-hyun** [comment]
Also, cc @williamhyun .

File: c++/include/orc/orc-config.hh.in (hunk 1 above)
Comment (wgtmac) [reply c++/include/orc/orc-config.hh.in:28]: > Just a question. Do we have any deprecation process in Apache ORC C++, @wgtmac , @stiga-huang ? I'm wondering when we can clean up these unused macros in the future.

IMHO, these kinds of breaking change (e.g. remove deprecated stuff) can be included in a major release. C++14 adds the support of `[[deprecated]]` and we can leverage it to deal with the change of public API. But it is tricky to deal with macros (e.g. add a new macro at the compile time to enable the deprecated macros).

In short, we should keep backward compatibility until the next major release (e.g. v2.0.0) and let the downstream know anything is going to be deprecated.

Comment (dongjoon-hyun) [reply c++/include/orc/orc-config.hh.in:28]: Got it. Thank you. I agree with you. We can do that later separately if needed.
